guppylang/prelude/_internal/compiler/array.py

```diff
@@ -12,229 +10,196 @@
+
+    def build_linear_array(self, elems: list[Wire]) -> Wire:
+        """Lowers a call to `array.__new__` for linear arrays."""
+        elem_opt_ty = ht.Option(self.elem_ty)
+        elem_opts = [
+            self.builder.add_op(ops.Tag(1, elem_opt_ty), elem) for elem in elems
+        ]
+        return self.builder.add_op(array_new(elem_opt_ty, len(elems)), *elem_opts)
+
     def compile(self, args: list[Wire]) -> list[Wire]:
-        match self.type_args:
-            case [TypeArg(ty=elem_ty), ConstArg(ConstValue(value=int(length)))]:
-                op = new_array(length, elem_ty.to_hugr())
-                return [self.builder.add_op(op, *args)]
-            case type_args:
-                raise InternalGuppyError(f"Invalid array type args: {type_args}")
+        if self.elem_ty.type_bound() == ht.TypeBound.Any:
+            return [self.build_linear_array(args)]
+        else:
+            return [self.build_classical_array(args)]
 
 
-class ArrayGetitemCompiler(CustomCallCompiler):
+class ArrayGetitemCompiler(ArrayCompiler):
     """Compiler for the `array.__getitem__` function."""
 
-    def build_classical_getitem(
-        self,
-        array: Wire,
-        array_ty: ht.Type,
-        idx: Wire,
-        idx_ty: ht.Type,
-        elem_ty: ht.Type,
-    ) -> CallReturnWires:
+    def build_classical_getitem(self, array: Wire, idx: Wire) -> CallReturnWires:
         """Lowers a call to `array.__getitem__` for classical arrays."""
-        length = self.type_args[1].to_hugr()
-        elem = build_array_get(
-            self.builder, array, array_ty, idx, idx_ty, elem_ty, length
-        )
+        idx = self.builder.add_op(convert_itousize(), idx)
+        result = self.builder.add_op(array_get(self.elem_ty, self.length), array, idx)
+        elem = build_unwrap(self.builder, result, "Array index out of bounds")
         return CallReturnWires(regular_returns=[elem], inout_returns=[array])
 
-    def build_linear_getitem(
-        self,
-        array: Wire,
-        array_ty: ht.Type,
-        idx: Wire,
-        idx_ty: ht.Type,
-        elem_ty: ht.Type,
-    ) -> CallReturnWires:
+    def build_linear_getitem(self, array: Wire, idx: Wire) -> CallReturnWires:
         """Lowers a call to `array.__getitem__` for linear arrays."""
         # Swap out the element at the given index with `None`. The `to_hugr`
         # implementation of the array type ensures that linear element types are turned
         # into optionals.
-        elem_opt_ty = ht.Sum([[elem_ty], []])
-        none = self.builder.add_op(ops.Tag(1, elem_opt_ty))
-        length = self.type_args[1].to_hugr()
-        array, elem_opt = build_array_set(
-            self.builder,
-            array,
-            array_ty,
-            idx,
-            idx_ty,
-            none,
-            elem_opt_ty,
-            length,
+        elem_opt_ty = ht.Option(self.elem_ty)
+        none = self.builder.add_op(ops.Tag(0, elem_opt_ty))
```

Comment:
maybe we should add `build_some` `build_none` `build_left` `build_ok` `build_error` `build_right` etc. to hugr-py?

Reply:
or maybe just constructors for the `Tag` op

Reply:
I think that's a good idea 👍 See https://github.com/CQCL/hugr/issues/1592